src/Vault.sol

```diff
@@ -0,0 +1,261 @@
+        // // shares are burned and after the assets are transferred, which is a valid state.
+        // _burn(owner, shares);
+        // SafeERC20.safeTransfer($._asset, receiver, assets);
+
+        // emit Withdraw(caller, receiver, owner, assets, shares);
+    }
+
+    // QUESTION: Start with Strategies or add them later
+    // vault starts paused because the rate provider and assets / strategies haven't been set
+    function initialize(address admin, string memory name, string memory symbol) public initializer {
+        // Initialize the vault
+        __ERC20_init(name, symbol);
+        __AccessControl_init();
+        __ReentrancyGuard_init();
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+
+        VaultStorage storage vaultStorage = Storage.getVaultStorage();
+        vaultStorage.paused = true;
+    }
+
+    // QUESTION: Measure the gas difference between IERC20 or address when casting / saving to storage
+    function setRateProvider(address rateProvider_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        VaultStorage storage vaultStorage = Storage.getVaultStorage();
+        vaultStorage.rateProvider = rateProvider_;
+        emit SetRateProvider(rateProvider_);
+    }
+
+    function addAsset(address asset_, uint8 decimals_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (asset_ == address(0)) revert ZeroAddress();
+        if (decimals_ > 18) revert InvalidDecimals();
+
+        AssetStorage storage assetStorage = Storage.getAssetStorage();
+
+        if (assetStorage.assets[asset_].decimals != 0) revert InvalidAsset();
+
+        uint256 newIndex = assetStorage.list.length;
+
+        assetStorage.assets[asset_] =
+            AssetParams({active: true, index: newIndex, decimals: decimals_, idleAssets: 0, deployedAssets: 0});
+
+        assetStorage.list.push(asset_);
+
+        emit AddAsset(asset_, decimals_, newIndex);
+    }
+
+    function toggleAsset(address asset_, bool active) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        AssetStorage storage assetStorage = Storage.getAssetStorage();
+        if (assetStorage.assets[asset_].decimals == 0) revert AssetNotFound();
+        assetStorage.assets[asset_].active = active;
+        emit ToggleAsset(asset_, active);
+    }
+
+    function pause(bool paused_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        VaultStorage storage vaultStorage = Storage.getVaultStorage();
+        vaultStorage.paused = paused_;
+        emit Pause(paused_);
+    }
+
+    // QUESTION: What params should be used here? strategy, asset, etc.
+    function processAccounting() public {
+        // get the balances of the assets
+        // AssetStorage storage assetStorage = _getAssetStorage();
+
+        // for (uint256 i = 0; i < assetsStorage.list.length; i++) {
+        //     address asset = assets[i];
+        //     idleBalance = asset.balanceOf(address(vault));
+        // }
+        // get the balances of the strategies
+
+        // call convertToAssets on the strategie?
```

Comment:
assuming the strategies are ERC20s this can actually simplify the logic here 

and push the logic of calculating the value of X amount of Shares of Strategy to the rate provider

Reply:
Another case here is if you're not dealing with a strategy that's an ERC20.

It can be an NFT (eg. a Uniswap V3 position)

A pending withdrawal is similar to that as well.

so potentially we'd have to have a TVL module here to customize for things that are not ERC20s